TLS enroll - send useful info with enroll request

lambda-conjecture opened the pull request:
I added two features. 
1. Added a tls_config_enroll_on_startup flag, defaulted to false. At default, there is no change in tls config behavior. If set to true, TLSConfigPlugin::setUp clears any cached node key and causes an enroll. This is useful for back-ends that do not want to cache node-keys for security reasons.
2. Added code to TLSEnrollPlugin so that on enroll, it sends osquery_info and os_version info to the backend. This gives the backend useful information that it can use to create the correct configurations for the node, amongst other things. 

I also added tests for TLSConfigPlugin, to verify both the default behavior as well as the behavior when tls_config_enroll_on_startup is set to true. I also modified test_http_server.py, so that it responds to a new endpoint, "/test", and responds with a cached list of requests it has received. This will allow unit tests to verify tls functionality.

osqueryer commented:
Can one of the osquery admins reply with "ok to test" to kick off a build...

osqueryer commented:
Can one of the osquery admins reply with "ok to test" to kick off a build...

osqueryer commented:
Can one of the osquery admins reply with "ok to test" to kick off a build...

mtmcgrew commented:
ok to test

osqueryer commented:
:no_good: The commit 69e58db29dbbcb31642491ce90b014e5ea5c7721 (Job results: [380](https://jenkins.osquery.io/job/osqueryPullRequestAudit/380/)) failed the code audit.

osqueryer commented:
:-1: The commit 69e58db29dbbcb31642491ce90b014e5ea5c7721 (Job results: [402](https://jenkins.osquery.io/job/osqueryPullRequestBuild/402/)) failed one or more tests (**Windows**).

osqueryer commented:
:no_good: The commit 4a0dee03970220b1e49a6a834781930f011a822c (Job results: [400](https://jenkins.osquery.io/job/osqueryPullRequestAudit/400/)) failed the code audit.

osqueryer commented:
:-1: The commit 4a0dee03970220b1e49a6a834781930f011a822c (Job results: [422](https://jenkins.osquery.io/job/osqueryPullRequestBuild/422/)) failed one or more tests (**Windows**).

osqueryer commented:
:no_good: The commit ea0ad8c8fabf01d6ef6d0f00e31d4a1447e73871 (Job results: [401](https://jenkins.osquery.io/job/osqueryPullRequestAudit/401/)) failed the code audit.

lambda-conjecture commented:
Updated the PR - removed os_version, since that table doesn't yet exist for windows. The build_platform and build_distro of osquery_info are sufficient to identify the host's os for most configuration purposes. Working on os_version table for windows for a subsequent PR.

PoppySeedPlehzr commented:
@lambda-conjecture PR #2586 contains the implementation for the os_version table for Windows, and it should hopefully be merged in relatively soon. Once that's done mind re-basing to pull in those changes?

lambda-conjecture commented:
Will do - waiting for PR #2586.

marpaia commented:
This is cool! @zwass and I were recently chatting about how we would love to get some info from the client on enrollment, but it also included hostname, ip address, etc. It would be cool if, instead of hard-coding the results of `osquery_info` into the enrollment request, there was an option to declare additional parameters in the form of osquery queries. Perhaps an additional "enroll" key in the decorators config parser? For example, the following would have the same effect as this PR:

```json
  "decorators": {
    "enroll": [
      "SELECT * FROM osquery_info LIMIT 1;"
    ],
    "load": [
      "SELECT uuid AS host_uuid FROM system_info;",
      "SELECT user AS username FROM logged_in_users ORDER BY time DESC LIMIT 1;"
    ]
  },
```

Thoughts?

lambda-conjecture commented:
It would definitely be a very elegant solution to use decorators. That said, I think it becomes a trade-off between a deterministic/simple bootstrap protocol  vs reuse of (an elegant piece of) machinery. The catch-22 I was running into is that, in order to send the correct config, the protocol must be some variant of:

osqueryd              backend

config ------------->
           <------------ invalid-node
enroll -------------->
           <------------ node-key
distributed_read-->
           <------------ "select * from <whatever_tables_I_need>
distributed_write-->
                             record basic info
config ---------------> 
           <------------- valid config


Using decorators, it would work as follows

config --------------->
           <--------------  Send "bootstrap config" with decorators
any-request -------> 
           <-------------- Invalid node key (to force an enroll)
enroll ----------------> Record basic info via decorator query response
           <-------------- Node Key
config ----------------> 
           <-------------- Correct config based on enrollment data

With my PR request, the bootstrap protocol is

enroll ----------------> Has enough data for bootstrap (it already has host uuid or host name)
          <--------------- Node Key
config ---------------> 
          <--------------- Correct config based on enrollment data

Putting it all together, perhaps a better idea is to read in the enroll "decorator" queries from a json file passed in via command line. Then, you can have the best of all worlds, although, at that point, we are mutating the elegant implementation of decorators (ie., they always come from configs).

marpaia commented:
It's a good point that in order for a client to know what decorator queries to send, it must already be enrolled and have received a valid config. That makes it kind of a silly solution IMO.

I think having to drop another file on the system for this case is also not super awesome, although it's similar in design to the `extensions_autoload` flag: https://github.com/facebook/osquery/blob/master/docs/wiki/deployment/extensions.md#autoloading-extensions

I think a `\n` delimited file would be more consistent than JSON for that reason.

I could also see just adding a flag which was a single SQL query. Something like `--enroll_metadata="select * from osquery_info limit 1"`.

I personally like the `--enroll_metadata` option or something, as long as it doesn't require me to write and manage yet another file to disk.

zwass commented:
I like the idea of the `--enroll_metadata` flag. It could default to `select * from osquery_info limit 1`, but would provide a ton of flexibility so that multiple tables could be selected from, etc.

lambda-conjecture commented:
LOL - I became really enamored of the decorators solution Mike had originally suggested (The idea of decorators is elegant, and intuitively, what we are really trying to do is decorate an enroll request). I see a way to make that work and still cover all the scenarios. The idea is to allow enroll to be its own source of decorators (source=enroll) for decorations of a new DecorationPoint type, DECORATE_ENROLL. Enroll requests will then have the appropriate decorations. The nice thing here is that configuration plugins can also insert their own DECORATE_ENROLL decorations (source=plugin-name), and the corresponding enroll plugins will pick up those decorations.

The source=enroll decorations will come from the command line as suggested by zwass and mike above. I was thinking of adding the ability to also pick them up from a json file. 

This solution covers all the situations:
1. Default behavior remains unchanged from what it is today
2. Easy way to add a simple query as a decorator through command line invocation flag
3. Completely flexible way to add enroll decorations through a JSON file
4. Ability to add/update/delete enroll decorations remotely

Internally, we use the existing decorator mechanism to store/update/run the decorator queries, so the implementation should be cohesive with the current design pattern.

Too complex?

marpaia commented:
I'm not sure that I follow what the command line flag would be to declare decorators? Or are you saying that decorators should be able to be extended via environment variables? 

And my vote would be for a `/n` delimited file similar to `extensions_autoload` if we must add a new file, but I wouldn't want to use that personally.

lambda-conjecture commented:
The command line flag would be as suggested, a simple 
--enroll_decorators='select * from osquery_info LIMIT 1; select * from os_version;'.
Its just that internally, I would store this as a decorator, so I can leverage the existing decorator mechanism to store and run the queries, as well as decorate the enroll request.

The file mechanism would be an additional option, something like
--enroll_decorators_json (or, if you prefer, --enroll_decorators_file, a \n delimited file), in case there is a need to describe a complex query or some more extensive set of queries.
Perhaps I can leave this option for a future date, when there is an expressed need for it.

As for the config having enroll decorators, the use-case I can think of is that perhaps the back-end wants to see certain info on a re-enrollment in order to more quickly re-authenticate the host. If I use the decorator mechanism for storing/running/decorating the command-line supplied queries, the config option sort of falls out almost for free.

marpaia commented:
> The command line flag would be as suggested, a simple 
--enroll_decorators='select * from osquery_info LIMIT 1; select * from os_version;'.
Its just that internally, I would store this as a decorator, so I can leverage the existing decorator mechanism to store and run the queries, as well as decorate the enroll request.

I see, that's cool. But what if a config get's delivered that has it's own decorators? Would they get merged in some way? Overwritten?

> The file mechanism would be an additional option, something like
--enroll_decorators_json (or, if you prefer, --enroll_decorators_file, a \n delimited file), in case there is a need to describe a complex query or some more extensive set of queries.
Perhaps I can leave this option for a future date, when there is an expressed need for it.

I see, yeah, if you really want this, then cool, but I think it would be super reasonable to leave this to future work if it becomes necessary for your use-case.

> As for the config having enroll decorators, the use-case I can think of is that perhaps the back-end wants to see certain info on a re-enrollment in order to more quickly re-authenticate the host. If I use the decorator mechanism for storing/running/decorating the command-line supplied queries, the config option sort of falls out almost for free.

But if you tell a client to re-enroll, does it purge the config? What if you've told the client to re-enroll because you want to send it new enroll decorator queries? I'm just not sure about how the semantics of using things in the config at a time when you're not supposed to have a config would work, but perhaps that's a conversation better had over an implementation PR?

theopolis commented:
There are two changes in the PR that may be better discussed if proposed separately.

1. I don't understand the node-key regeneration on startup concept. The node-key is designed to identify/authenticate a host to your backend. If you're not using it for host authentication consider a TLS client authentication model.

I don't understand the premise "This is useful for back-ends that do not want to cache node-keys for security reasons", mind explaining a bit more?

2. The enrollment should not be overloaded into a general purpose "give me details" step in my opinion. It seems like some folks want course details about the system to aide in inventorying and identification. Is it possible to simplify this implementation by ONLY returning system-identification information?

The proposed: `osquery_info`, `os_version` tables are OK. It seems fine to be flexible in appending data to the enroll step without changing APIs/versions/flags. I might recommend `system_info` too.

Keep in mind, you can get this information in other ways now. For our deployments, upon enrollment, several queries are added to the schedule and issues as ad-hoc queries to populate this information.

theopolis commented on osquery/config/plugins/tests/tls_config_tests.cpp:
```diff
@@ -0,0 +1,175 @@
+class TLSConfigTests : public testing::Test {};
+
+TEST_F(TLSConfigTests, test_setup) {
+  auto tlsConfigPlugin = Registry::get("config", "tls");
```
Unless you're writing Windows code, prefer snake case please :)

lambda-conjecture replied:
That obvious, eh? :-)

theopolis commented on osquery/config/plugins/tests/tls_config_tests.cpp:
```diff
@@ -0,0 +1,175 @@
+TEST_F(TLSConfigTests, test_setup) {
+  auto tlsConfigPlugin = Registry::get("config", "tls");
+
+  std::string uri;
```
Move these closer to their use, it will help with scope limiting later.

theopolis commented on osquery/config/plugins/tests/tls_config_tests.cpp:
```diff
@@ -0,0 +1,175 @@
+  status = setDatabaseValue(kPersistentSettings, "nodeKeyTime", dbValue);
+  ASSERT_TRUE(status.ok());
+
+  // --------------------------------------------------------------------------
```
This is weird, please do not use section/header/etc meta-styling-comments.

theopolis commented on osquery/config/plugins/tests/tls_config_tests.cpp:
```diff
@@ -0,0 +1,175 @@
+
+class TLSConfigTests : public testing::Test {};
+
+TEST_F(TLSConfigTests, test_setup) {
```
I'm not really sure what additional code is tested here or the goal.

Enroll workflow tests: https://github.com/facebook/osquery/blob/master/osquery/remote/enroll/tests/enroll_tests.cpp

Config workflow tests:
https://github.com/facebook/osquery/blob/master/osquery/config/tests/config_tests.cpp

Remote workflow tests:
https://github.com/facebook/osquery/blob/master/osquery/remote/transports/tests/tls_transports_tests.cpp

If you want to test the additional data sent via enroll, let's prefer to capture it within the abstract enroll tests.

lambda-conjecture replied:
The main motivation of adding the tls_config_tests was to test the behavior of the config_tls_enroll_on_startup flag, which was newly added as part of this PR

theopolis commented on tools/tests/test_http_server.py:
```diff
@@ -163,6 +168,14 @@ def distributed_write(self, request):
     def log(self, request):
         self._reply({})
 
+    def pushrequest(self, command, request):
```
nit, use snake_case please.

theopolis commented on tools/tests/test_http_server.py:
```diff
@@ -163,6 +168,14 @@ def distributed_write(self, request):
+        request['command'] = command
+        RECEIVED_REQUESTS.append(request)
+        
+    def test(self):
```
nit, let's make this more meaningful.

theopolis commented on osquery/remote/enroll/plugins/tls.cpp:
```diff
@@ -20,6 +22,12 @@
 
 namespace osquery {
+namespace tables {
+QueryData genOsqueryInfo(QueryContext& context);
```
If you see this style of "I need to use a table" anywhere in the code base please fix it, it's poor design but we do use it in a few cases.

We should avoid "bind" tables to "core" osquery components. Instead prefer to use the `SQL` class.

lambda-conjecture replied:
The main consideration here was making assumptions about how much of the system can be expected to be initialized at the time this code is likely to be called. If the "config_tls_enroll_on_startup" option is set, then enroll gets called at the time that the active config plugin gets initialized (osquery/core/init.cpp, line~587). This is *before* the Registry::setup() is called. I must confess that I do not have knowledge of the designers' full (future) intentions here, but I was driven by the comment "Run the setup for all lazy registries (tables, SQL)" describing the call to Registry::setUp() to be defensive and chose to call the genXXX queries directly.

Happy to switch to using the SQL class

theopolis commented on osquery/remote/enroll/plugins/tls.cpp:
```diff
@@ -87,6 +95,15 @@ Status TLSEnrollPlugin::requestKey(const std::string& uri,
 
+  QueryContext qc;
+
+  auto osqueryInfo = tables::genOsqueryInfo(qc);
```
Replace with:
```cpp
auto results = SQL::selectAllFrom("osquery_info");
```

marpaia commented:
@theopolis i think that having the enroll request contain details about the host is useful. Many osquery backends, including the one i'm working on, have a `hosts` database table that might look something like:

```
CREATE TABLE `hosts` (
  `id` int(10) unsigned NOT NULL AUTO_INCREMENT,
  `node_key` varchar(255) DEFAULT NULL,
  `host_name` varchar(255) DEFAULT NULL,
  `uuid` varchar(255) DEFAULT NULL,
  `ip_address` varchar(255) DEFAULT NULL,
  `platform` varchar(255) DEFAULT NULL,
  `osquery_version` varchar(255) DEFAULT NULL,
  // arbitrary other columns
  PRIMARY KEY (`id`)
) ENGINE=InnoDB DEFAULT CHARSET=latin1
```

Right now, the process for enrolling a host is similar to how you describe it. A host enrolls and we create a new records in `hosts`. Then, we have a set of distributed queries which we execute to populate the details in hosts table. Unfortunately, however, while we're waiting for the host to ask for, execute, and submit those distributed queries, we don't know anything about the host. 

This isn't a problem really if you consider that when you have tens or hundreds of thousands of hosts, you're probably not paying attention to them during that time period or with any level of individuality. Most companies, however, have far less hosts. This makes it noticeable when a new host enrolls but you don't know anything about it yet. We could handle some logic to hide enrolled hosts that have not submitted their detail queries yet, but that's just unnecessary.

Consider also that if you had a UX which displayed hosts. If the only data you have about hosts is coming from osquery, not knowing anything about a subset of hosts for any amount of time is a poor experience. 

A configurable way to optionally send host metadata along with enrollment would unequivocally solve this problem though, which is why I support it.

lambda-conjecture commented:
All, thanks for the feedback - much appreciated. Working on incorporating most of the ideas.

lambda-conjecture commented:
@theopolis, wrt your comment "I'm not really sure what additional code is tested here or the goal."

I added a new flag/option to force an enroll request before the first config, for almost exactly the use-cases @theopolis described - briefly, to have a backend be more responsive to newly enrolled hosts. The tests I added were to ensure that config::setup works as expected with and without the flag turned on. The test verifies that with the flag defaulted, there is no change in behavior, and then verifies that with the flag turned on, config infact forces an enroll before sending the config.

osqueryer commented:
:no_good: The commit baef13d23bb61258cb645a4df841ec0603e01483 (Job results: [550](https://jenkins.osquery.io/job/osqueryPullRequestAudit/550/)) failed the code audit.

osqueryer commented:
:-1: The commit baef13d23bb61258cb645a4df841ec0603e01483 (Job results: [3568](https://jenkins.osquery.io/job/osqueryPullRequestBuild/3568/)) failed one or more tests (**OS X/Linux**).

osqueryer commented:
:-1: The commit 9abbc7e0ce82162fe08cddc2f25c9162ec1a0187 (Job results: [575](https://jenkins.osquery.io/job/osqueryPullRequestBuild/575/)) failed one or more tests (**Windows**).

osqueryer commented:
:-1: The commit 9abbc7e0ce82162fe08cddc2f25c9162ec1a0187 (Job results: [3569](https://jenkins.osquery.io/job/osqueryPullRequestBuild/3569/)) failed one or more tests (**OS X/Linux**).

lambda-conjecture commented:
I updated this PR with the following:

1. Per requests above, I added a command line option, "--enroll_tls_metadata", which takes a string argument. The string format is "name:query;name:query;", for example, "osquery_info:select * from osquery_info;ip_address:select * from interface_addresses;". If specified, the enroll request has an array of objects named "name", each one representing a row of the SQL query results. In the example above, the enroll request JSON will have one object called "osquery_info" because select * from osquery_info only returns 1 row. It will have an array of "ip_address" objects, each one with the fields of a single row of the interface_addresses table.

2. I have incorporated all (I hope) of the reviewers comments wrt putting code closer to it's logical home, snake_case, more meaningful names for some of the test functions, etc. 

3. I experimented with using decorators. While I got it to work, I abandoned this implementation because I had to twist too many design elements of decorators. For example, the decoration point refers to the evaluation point of the query; regardless of decoration point, all the query results are added to each log element. For enroll, I had to split the "store" of decorator query results into two, one for log elements and one for enroll. Also, a useful case is to return all the IP addresses of a machine, which of course, is a multi-row SQL query result. Decorators are designed to return single-row query results. The finished code using decorators required too many changes, and the results were not entirely satisfactory (for example, the only way to return multi-row results were to add a sequence number to the field names etc., basically a hack).

lambda-conjecture commented:
PS: The checks below are failing for what I believe are environmental reasons - the centos-7 build failed to pull down the python package, and the windows build threw a java exception inside Jenkins. Can someone please requeue the jobs?

theopolis commented:
ok to test

osqueryer commented:
:no_good: The commit a29e2f9b743f885317c7412b47ab1a9abfd87f2f (Job results: [712](https://jenkins.osquery.io/job/osqueryPullRequestBuild/712/)) failed the code audit or documentation test.

osqueryer commented:
:-1: The commit a29e2f9b743f885317c7412b47ab1a9abfd87f2f (Job results: [728](https://jenkins.osquery.io/job/osqueryPullRequestBuild/728/)) failed one or more tests (**Windows**).

theopolis commented on osquery/config/plugins/tests/tls_config_tests.cpp:
```diff
@@ -54,7 +73,84 @@ TEST_F(TLSConfigTests, test_retrieve_config) {
+  ASSERT_TRUE(status.ok());
+
+  db_value = std::to_string(getUnixTime());
+  status = setDatabaseValue(kPersistentSettings, "nodeKeyTime", db_value);
```
Are you using this? I see this is in tests from before, but it seems unused.

theopolis commented on tools/tests/test_http_server.py:
```diff
@@ -178,6 +183,16 @@ def distributed_write(self, request):
         self._reply({})
 
+    def test_read_requests(self):
+        # call made by unit tests to retrieve the last request made by
```
nit, wording, looks like a thought wasn't completed?

theopolis commented on osquery/remote/enroll/plugins/tls.cpp:
```diff
@@ -12,13 +12,20 @@
+#include "osquery/core/conversions.h"
 #include "osquery/core/process.h"
 
+#include "tls.h"
```
Use the complete local path from the source directory please. :)

theopolis commented on osquery/remote/CMakeLists.txt:
```diff
@@ -9,7 +9,7 @@ ADD_OSQUERY_LIBRARY(FALSE osquery_enrollment_plugins
 )
 
-file(GLOB OSQUERY_ENROLLMENT_PLUGIN_TESTS "enrollment/plugins/tests/*.cpp")
+file(GLOB OSQUERY_ENROLLMENT_PLUGIN_TESTS "enroll/plugins/tests/*.cpp")
```
Nice catch!

theopolis commented on osquery/config/plugins/tls.cpp:
```diff
@@ -46,32 +48,33 @@ CLI_FLAG(uint64,
 REGISTER(TLSConfigPlugin, "config", "tls");
 
 Status TLSConfigPlugin::setUp() {
+  if (FLAGS_config_tls_enroll_on_startup && !FLAGS_disable_enrollment) {
```
I'd like to disconnect this "ephemeral" enroll concept from the specific TLS config and TLS enroll plugins if it's not too much work?

theopolis replied:
Maybe we call this `--enroll_ephemeral` and it always deletes the host UUID and cached node key on start.

theopolis commented on osquery/config/plugins/tls.cpp:
```diff
@@ -25,6 +25,8 @@
 #include "osquery/remote/serializers/json.h"
 #include "osquery/remote/utility.h"
 
+#include "tls.h"
```
nit, please use the complete local path from the base of the project.

theopolis commented on osquery/config/plugins/tests/tls_config_tests.cpp:
```diff
@@ -9,17 +9,36 @@
  */
 
 #include <gtest/gtest.h>
+#include <vector>
```
Place these C++ includes above our local dependencies include, and put a space between the groupings please.

theopolis commented:
I still think we should break apart the two features here:
- The ephemeral enrollment concepts
- The enrollment metadata

The ephemeral, always re-enroll, is looking good. With a few more style tweaks I think it's ready to land. The enrollment metadata concept still needs more discussion and will block this PR from merging for a while unfortunately.

osqueryer commented:
:-1: The commit 21da2ae1396ebc3546b79ef0b0d82654ea05a09e (Job results: [793](https://jenkins.osquery.io/job/osqueryPullRequestWindows/793/)) failed one or more tests (**Windows**).

osqueryer commented:
:-1: The commit f70695ba9df3129029bb2552373f84956c4287db (Job results: [919](https://jenkins.osquery.io/job/osqueryPullRequestWindows/919/)) failed one or more tests (**Windows**).

lambda-conjecture commented:
So, this PR has been trimmed to just adding the enroll_tls_metadata flag. This flag allows you to specify queries whose results will be sent as metadata along with an enroll request. This gives TLS backends a way to get information about the enrolling host, enabling it to drive the rest of the interaction with the host. For example, one can use this to send the output of "select * from os_version", which might enable the backend to select the set of query-packs to send to the newly enrolled host.

lambda-conjecture commented:
Please let me know if anyone needs more info to resume a conversation about this proposed feature.

theopolis commented:
This is on my short list of significant features that need review, sorry for the delay.

theopolis commented:
Ok, circling back to this. Is there a short-list of things to send when enrolling? Can we make this a little less flexible, aka no queries as flagfiles? I think the limit of flexibility would be a CSV of table names within a flag.

osqueryer commented:
:-1: The commit 023af89d4c2bb4cabdad721b1902ae47e48c8541 (Job results: [177](https://jenkins.osquery.io/job/osqueryPullRequestLinux/177/)) failed one or more tests (**FreeBSD**).

marpaia commented:
If we were limiting this to a shortlist, I would propose including the following single-row tables:
- `os_version`
- `platform_info`
- `osquery_info`
- `system_info`

thedrow commented:
Any news on this one?

theopolis commented:
I'll begin work on this shortly, it still makes sense and is med-pri.
